### PR TITLE
S3 compatible Foundry API, boto methods and aws cli profile

### PR DESCRIPTION
**Pull request by jonas-w**
# Summary
Makes it easier to get started with the new beta s3 dataset API.

# Checklist

- [x] You agree with our [CLA](https://gist.githubusercontent.com/emdgroup-admin/16cc45ea4315c2ef29eb9d9afc36fcf5/raw/abb9c91f15278a62b9ac3e66144bcd27fa9485c2/CLA.md)
- [x] Included tests (or is not applicable).
- [x] Updated [documentation](https://emdgroup.github.io/foundry-dev-tools/) (or is not applicable).
- [x] Used [pre-commit hooks](https://emdgroup.github.io/foundry-dev-tools/develop.html#pre-commit-hooks-formatting) to format and lint the code.

### CODE REVIEW

File: docs/s3.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+# -------------------
+
+fs = fsspec.filesystem("s3", **fc.get_s3fs_storage_options())
+with fs.open(f"{dataset_rid}/fsspec_test_write.txt", "w") as f:
+    f.write("hihi")
+with fs.open(f"{dataset_rid}/fsspec_test_write.txt", "r") as f:
+    print(f.read())
+
+````
```
Comment (nicornk) [inline docs/s3.md:108]: I think docs not an update after recent changes.

File: docs/s3.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+fdt s3 auth
+```
+
+The auth command gets used by the configuration profile as mentioned above, it adheres to the output schema defined at https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sourcing-external.html
+
+
+## Methods you can use in python
+
+- [FoundryRestClient.get_s3_client](#foundry_dev_tools.foundry_api_client.FoundryRestClient.get_s3_client)
+- [FoundryRestClient.get_s3_resource](#foundry_dev_tools.foundry_api_client.FoundryRestClient.get_s3_resource)
```
Comment (nicornk) [inline docs/s3.md]: Let’s allow user to pass config or additional kwargs, for example to set retry settings https://stackoverflow.com/a/48568320

Comment (nicornk) [inline docs/s3.md]: fc.get_boto3_client()
fc.get_boto3_resource()
Like those more

File: docs/s3.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+- [FoundryRestClient.get_s3fs_storage_options](#foundry_dev_tools.foundry_api_client.FoundryRestClient.get_s3fs_storage_options)
+- [FoundryRestClient.get_s3_polars_storage_options](#foundry_dev_tools.foundry_api_client.FoundryRestClient.get_s3_polars_storage_options)
+
```
Comment (nicornk) [inline docs/s3.md:42]: Let’s remove polars for now, since it seems to not work.

File: src/foundry_dev_tools/cli/s3.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    with aws_config_file.open("w+") as aws_config_fd:
+        aws_config_fd.write(new_config)
+
+
+@click.command("auth")
+def s3_cli_auth():
+    """Used by the aws cli for authentication."""
+    # print everything that happens when initalizing i.e. oath input() prompt
+    # to stderr, which gets redirected to the console through the hacks in the profile config
+    orig = sys.stdout
```
Comment (nicornk) [inline src/foundry_dev_tools/cli/s3.py:104]: Hack -> workaround 
Can we add the AWS Issue (that asks for better support) in the comment

File: src/foundry_dev_tools/foundry_api_client.py [hunk 5]
```diff
@@ -2005,6 +1995,124 @@ def _query_fsql(
             return arrow_stream_to_spark_dataframe(arrow_stream_reader)
         return arrow_stream_reader.read_all()
 
+    def get_s3_credentials(self) -> dict:
```
Comment (nicornk) [inline src/foundry_dev_tools/foundry_api_client.py]: We might want to expose the expiration in minutes as argument. I think the api supports it.

File: src/foundry_dev_tools/foundry_api_client.py [hunk 6]
```diff
@@ -2005,6 +1995,124 @@ def _query_fsql(
...
+        """
+        t = self._requests_session.post(
+            self._s3_url,
+            params={
+                "Action": "AssumeRoleWithWebIdentity",
+                "WebIdentityToken": _get_auth_token(self._config),
+            },
+        ).text
+        return {
+            "access_key": t[
```
Comment (nicornk) [inline src/foundry_dev_tools/foundry_api_client.py]: Pull the response transformation into a private method and unit test it.

File: src/foundry_dev_tools/utils/async_s3.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        self,
+        foundry_rest_client: FoundryRestClient,
+        session: aiobotocore.session.Session | None = None,
+    ):
+        self.foundry_rest_client = foundry_rest_client
+        super().__init__(session)
+
+    async def load(self):
+        """Return the credentials from FoundryRestClient."""
+        # Here you should implement your logic to get the credentials.
```
Comment (nicornk) [inline src/foundry_dev_tools/utils/async_s3.py]: Remove comment

File: docs/s3.md (hunk 2 above)
Comment (jonas-w) [reply docs/s3.md]: @nicornk but these are s3 clients/resources, not 'boto3 clients/resources', maybe could be changed to `get_boto3_s3_client/resource`

File: src/foundry_dev_tools/foundry_api_client.py (hunk 6 above)
Comment (jonas-w) [reply src/foundry_dev_tools/foundry_api_client.py]: now using `defusedxml` for parsing the response.

**Pull request merged**